Admin: add schema field to resource

AntoineAugusti opened the pull request:
![image](https://user-images.githubusercontent.com/295709/87434566-a4eef700-c5b8-11ea-9918-27897e6f6ae0.png)

- Link data.gouv.fr to the schema catalog of schema.data.gouv.fr
- Add a schema field to the resource model
- Let producers select a schema from a list on resources

I left some comments/questions at various places.

## To discuss
- Configuration for the schema catalog URL
- The schema catalog of schema.data.gouv.fr only lists TableSchema schemas for now, should we change it? It's because we needed this endpoint for Validata, therefore it's made to include only TableSchema schemas. This can change.

AntoineAugusti commented on js/models/schemas.js:
```diff
@@ -0,0 +1,14 @@
+        super(options);
+        this.$options.ns = 'datasets';
+        this.$options.fetch = 'schemas';
+        this.sorted = 'label';
```
Sort schemas in alphabetical order for the frontend

AntoineAugusti commented on udata/core/dataset/api.py:
```diff
@@ -550,3 +554,13 @@ def get(self):
+class SchemasAPI(API):
+    @api.doc('schemas')
+    @api.marshal_list_with(schema_fields)
+    @cache.cached(timeout=60*5)
```
I wanted to add a configuration to specify a different TTL but I didn't manage to. It seems `current_app.config` isn't available here, not in the app context.

Any way to work around this? Is it needed?

abulte replied:
How about `SCHEMA_CACHE_DURATION = current_app.config.get()` at the module level and then use `SCHEMA_CACHE_DURATION`? I think it should work. https://github.com/opendatateam/udata/blob/master/udata/core/dataset/preview.py#L58

AntoineAugusti commented on udata/core/dataset/forms.py:
```diff
@@ -79,6 +80,10 @@ class BaseResourceForm(ModelForm):
     extras = fields.ExtrasField()
+    schema = fields.StringField(
+        _('Schema'),
+        # choices=[schema['id'] for schema in ResourceSchemas.get()],
```
Can't add this line here as `ResourceSchemas.get()` needs `current_app.config`. It isn't available there as it's not in the app context when called from here.

How to get around this?

abulte replied:
same as above, build the choices list at the module level.

AntoineAugusti commented on udata/core/dataset/models.py:
```diff
@@ -747,6 +750,23 @@ def from_community(self):
 
+class ResourceSchemas(object):
+    @staticmethod
+    def get():
```
No caching for this method, caching is done at the API level

AntoineAugusti commented on udata/settings.py:
```diff
@@ -437,6 +437,10 @@ class Defaults(object):
 
+    # Schemas parameters
+    ####################
+    SCHEMA_CATALOG_URL = None
```
Is it a sensible default? Should I change this?

AntoineAugusti commented on udata/tests/api/test_datasets_api.py:
```diff
@@ -1325,3 +1326,27 @@ def test_resource_types_list(self):
+
+
+@pytest.mark.usefixtures('clean_db')
+class DatasetSchemasAPITest:
```
Wrote tests for my new endpoint using the `pytest` syntax. Other classes inherited from `APITestCase`, not compatible with `pytest`.

AntoineAugusti commented on udata/tests/api/test_datasets_api.py:
```diff
@@ -1325,3 +1326,27 @@ def test_resource_types_list(self):
+
+@pytest.mark.usefixtures('clean_db')
+class DatasetSchemasAPITest:
+    modules = ['core.dataset']
```
I discovered adding this was key to get the routing of `api.get` working, can you explain why?

abulte replied:
It make the dataset module (and thus it's API) available for the test class. By default we try not to load too much stuff, in order to have faster tests.

AntoineAugusti commented on js/components/dataset/resource/form.vue:
```diff
@@ -243,7 +244,23 @@ export default {
         },
         fields() {
-            return this.generic_fields.concat(this.file_fields);
+            return this.generic_fields.concat(this.schema_field).concat(this.file_fields);
```
The schema field will not be added to the form if the schema catalog is empty / is not set.

AntoineAugusti commented on udata/core/dataset/search.py:
```diff
@@ -83,7 +83,8 @@ class Meta:
     resources = Object(properties={
```
Commenting for the whole file.

I'm not sure to understand how this class works. I've tried to look at various fields and added the necessary lines. 
I don't understand if I need to add (in order of changes, top to bottom):
- a property
- a facet
- update the `serialize` method

abulte replied:
I'm not sure either :-/ let's keep this for a future PR.

abulte commented on udata/core/dataset/models.py:
```diff
@@ -747,6 +750,23 @@ def from_community(self):
+
+        result = []
+        for schema in r.json()['schemas']:
+            result.append({'id': schema['name'], 'label': schema['title']})
```
I guess a list comprehension would work here

abulte commented on udata/tests/api/test_datasets_api.py:
```diff
@@ -1325,3 +1326,27 @@ def test_resource_types_list(self):
+    modules = ['core.dataset']
+
+    def test_dataset_schemas_api_list(self, app, api, rmock):
+        app.config['SCHEMA_CATALOG_URL'] = 'https://example.com/schemas'
```
use `@pytest.mark.options` decorator for this

abulte commented on udata/tests/dataset/test_dataset_model.py:
```diff
@@ -413,3 +415,18 @@ def test_multiple_strings(self):
+
+class ResourceSchemasTest:
+    def test_resource_schemas_get(self, app, rmock):
+        app.config['SCHEMA_CATALOG_URL'] = 'https://example.com/schemas'
```
use decorator, same as above

abulte commented on udata/tests/dataset/test_dataset_model.py:
```diff
@@ -413,3 +415,18 @@ def test_multiple_strings(self):
+        assert ResourceSchemas.get() == [{"id": "etalab/schema-irve", "label": "Schéma IRVE"}]
+
+    def test_resource_schemas_get_no_catalog_url(self, app):
+        app.config['SCHEMA_CATALOG_URL'] = None
```
use decorator

abulte commented on udata/tests/api/test_datasets_api.py:
```diff
@@ -1325,3 +1326,27 @@ def test_resource_types_list(self):
+        assert response.json == [{"id": "etalab/schema-irve", "label": "Schéma IRVE"}]
+
+    def test_dataset_schemas_api_list_no_catalog_url(self, app, api):
+        app.config['SCHEMA_CATALOG_URL'] = None
```
use decorator

abulte commented on udata/core/dataset/models.py:
```diff
@@ -747,6 +750,23 @@ def from_community(self):
+            return []
+
+        r = requests.get(endpoint)
+        r.raise_for_status()
```
I think we need a smarter caching strategy and exception handling.

Something like that https://github.com/etalab/udata-gouvfr/pull/486/files#diff-1c9abf3241dab91b0e60c554f8bcd67aR72.

This would lead to removing the cache on the API and handle it here.

abulte commented on udata/core/dataset/models.py:
```diff
@@ -747,6 +750,23 @@ def from_community(self):
         return True
 
 
+class ResourceSchemas(object):
```
our model classes are usually singular -> ResourceSchema

abulte commented:
> The schema catalog of schema.data.gouv.fr only lists TableSchema schemas for now, should we change it? It's because we needed this endpoint for Validata, therefore it's made to include only TableSchema schemas. This can change.

No need to change that for now. I don't think we'll do anything very interesting with other schemas for quite some time.

AntoineAugusti commented on udata/core/dataset/forms.py:
```diff
@@ -79,6 +80,12 @@ class BaseResourceForm(ModelForm):
     extras = fields.ExtrasField()
+    schema = fields.SelectField(
+        _('Schema'),
+        choices=[schema['id'] for schema in ResourceSchema.objects()],
```
Do I need to explicitly add a `None` to the list? Tests pass without it.

abulte replied:
No I don't think so.

AntoineAugusti commented on udata/core/dataset/api.py:
```diff
@@ -550,3 +551,18 @@ def get(self):
+        try:
+            # This method call is cached as it makes HTTP requests
+            return ResourceSchema.objects()
+        except LookupError:
+            abort(503, description='No schemas in cache and endpoint unavailable')
+        except ValueError:
```
It was difficult to pick appropriate exceptions to raise, I'm open to suggestions

abulte replied:
I wouldn't try to use base exceptions that are not made for that.

Either:
- use custom exceptions
- use the return value of the function (`[]` vs `None` for example) if it's enough

AntoineAugusti replied:
I added custom exceptions 🔧

AntoineAugusti commented on udata/core/dataset/models.py:
```diff
@@ -747,6 +756,45 @@ def from_community(self):
+class ResourceSchema(object):
+    @staticmethod
+    @cache.memoize(timeout=SCHEMA_CACHE_DURATION)
+    def objects():
```
Heavily inspired by the static pages' method you linked me to

AntoineAugusti commented:
Thanks for your guidance and your help @abulte!

I:
- tackled your PR comments
- reverted the commit introducing indexing - it'll be done later
- added a double caching mechanism to fetch schemas from the schema catalog
- cached at the model level instead of the API level
- enforced an enumeration for the `schema` property at the form level

AntoineAugusti commented:
Added custom exceptions as you suggested @abulte. Thanks for your review.

abulte commented on udata/core/dataset/models.py:
```diff
@@ -747,6 +759,45 @@ def from_community(self):
+            log.exception(f'Error while getting schema catalog from {endpoint}')
+            content = cache.get(cache_key)
+        else:
+            schemas = response.json()['schemas']
```
```suggestion
            schemas = response.json().get('schemas', [])
```

AntoineAugusti commented:
I had to do an [extra commit](https://github.com/opendatateam/udata/pull/2512/commits/49ac4a4fa4163f053798ac286fe733a2c5a1211d) this morning to fix `dataset/forms.py`. When starting my local server, I had a stracktrace regarding the absence of the app context in this file, because of `choices=[schema['id'] for schema in ResourceSchema.objects()]` which needs the app context under the hood. I guess the app context is not available yet, hence we can't access the current config at this point.

I switched to a string field with a validator. The SelectField relies on a list of options, through the `choices` keyword. You can't leave it empty, that's why I switched to the String class.

This pattern is used elsewhere in the very same file
https://github.com/opendatateam/udata/blob/32b09032f3b60ef26ae105ccdbce4591dae7dc6a/udata/core/dataset/forms.py#L33-L48

I made sure the validator was working by editing the HTML
![image](https://user-images.githubusercontent.com/295709/87947465-5e464480-ca71-11ea-9225-e72617cee2b3.png)
